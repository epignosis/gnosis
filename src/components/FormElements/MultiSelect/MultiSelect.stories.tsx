import React, { useState } from "react";
import { Story } from "@storybook/react";
import MultiSelectComponent, { MultiSelectOption, MultiSelectProps } from "./MultiSelect";

export default {
  title: "components/Form Elements/Multi Select",
  component: MultiSelectComponent,
  args: {
    placeholder: "Multi select",
    size: "md",
    label: "Choose multiple values",
    altLabel: false,
    block: true,
    id: "multiSelectStory",
    className: "multiSelectStory",
    options: [
      {
        label: "All",
        value: "all",
      },
      {
        label: "In progress",
        value: "progress",
      },
      {
        label: "Completed",
        value: "completed",
      },
      {
        label: "Failed",
        value: "failed",
      },
    ],
  },
  argTypes: {
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
  },
  decorators: [
    (Story: Story): JSX.Element => (
      <div style={{ maxWidth: 400 }}>
        <Story />
      </div>
    ),
  ],
};

export const MultiSelect: Story<MultiSelectProps> = (args) => {
  const [state, setState] = useState<MultiSelectOption[]>();

  return (
    <MultiSelectComponent
      {...args}
      id="multiSelect"
      value={state}
      onChange={(selections): void => setState(selections as MultiSelectOption[])}
    />
  );
};
