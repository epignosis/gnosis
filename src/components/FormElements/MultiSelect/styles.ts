import { css, Theme, SerializedStyles } from "@emotion/react";
import { inputBaseStyles, inputContainerBaseStyles } from "../styles";
import { InputSize } from "@components/FormElements/Input/Input";

const heightDimensions = {
  md: "2.5rem",
  lg: "3rem",
};

export const multiSelectContainer = (
  { typeScaleSizes, formElements }: Theme,
  { isOpen, block, size }: { isOpen: boolean; block: boolean; size: InputSize },
): SerializedStyles => css`
  ${inputContainerBaseStyles({ block })}
  ${block && "width: 100%"};
  position: relative;

  ul {
    font-size: ${typeScaleSizes.md};
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    list-style: none;
    display: ${isOpen ? "flex" : "none"};
    flex-direction: column;
    background-color: ${formElements.input.background};
    padding: 0.875rem 0;
    border: 1px solid ${formElements.input.inputBorderColor};
    border-radius: 5px;
    margin: 0;
    z-index: 1;

    li {
      align-items: baseline;
      padding: 0.25rem 1rem;
      margin: 0;

      &:hover {
        background: ${formElements.multiSelect.hover};
      }
    }
  }

  .select-btn {
    ${inputBaseStyles(formElements, { block })};
    ${block && "width: 100%"};
    height: ${heightDimensions[size]};
    display: inline-flex;
    justify-content: space-between;
    align-items: center;
    text-align: left;

    &:hover,
    &:focus {
      border: 1px solid ${formElements.input.borderHoverColor};
    }

    svg {
      margin-left: 2.5rem;
    }
  }

  .content {
    display: inline-flex;
    justify-content: space-between;
    align-items: start;
    width: 100%;
    margin-bottom: 0.5rem;
    padding: 0 1rem;

    span {
      color: ${formElements.input.placeholderColor};
      font-style: italic;
    }

    .close-btn {
      padding: 0;
      border: 0;
      background: transparent;
    }
  }
`;

export const option = ({ formElements }: Theme): SerializedStyles => css`
  display: inline-flex;
  align-items: center;
  margin-bottom: 0.5rem;
  cursor: default;

  &:last-of-type {
    margin-bottom: 0;
  }

  &.selected {
    span::after {
      background-color: #004ea8;
    }
  }

  span {
    height: 16px;
    width: 16px;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    border: 1px solid ${formElements.input.inputBorderColor};
    margin-right: 0.875rem;

    &::after {
      content: " ";
      height: 10px;
      width: 10px;
      display: inline-block;
      background-color: transparent;
    }
  }
`;
