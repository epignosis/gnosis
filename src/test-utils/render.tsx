import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
