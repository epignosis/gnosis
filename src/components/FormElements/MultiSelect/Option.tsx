import React, { forwardRef, ForwardRefRenderFunction } from "react";
import classNames from "classnames";
import { option } from "./styles";

export type OptionProps = React.PropsWithRef<JSX.IntrinsicElements["li"]> & {
  isSelected: boolean;
  isHighlighted: boolean;
};

const Option: ForwardRefRenderFunction<HTMLLIElement, OptionProps> = (props, forwardedRef) => {
  const { isSelected, isHighlighted, children, ...rest } = props;
  const containerClassNames = classNames({
    highlighted: isHighlighted,
    selected: isSelected,
  });

  return (
    <li ref={forwardedRef} css={option} className={containerClassNames} {...rest}>
      <span /> {children}
    </li>
  );
};

export default forwardRef(Option);
