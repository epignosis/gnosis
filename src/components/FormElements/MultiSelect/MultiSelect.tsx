import React, { FC } from "react";
import { useSelect, useMultipleSelection } from "downshift";
import { SerializedStyles } from "@emotion/react";
import classNames from "classnames";
import { multiSelectContainer } from "./styles";
import Option from "./Option";
import { Label, Text } from "@components";
import { InputSize } from "@components/FormElements/Input/Input";
import { CaretDownSVG, DropUpArrowSVG } from "@icons/core";

export type MultiSelectOption = {
  value: string;
  label: string;
};

export type MultiSelectProps = {
  placeholder: string;
  id?: string;
  size?: InputSize;
  label?: string;
  altLabel?: boolean;
  options: MultiSelectOption[];
  onChange: (selections: unknown[]) => void;
  value?: MultiSelectOption[] | undefined;
  className?: string;
  block?: boolean;
  children?: never;
};

const MultiSelect: FC<MultiSelectProps> = (props) => {
  const {
    placeholder,
    size = "md",
    id,
    options,
    onChange,
    value,
    className,
    block = false,
    label,
    altLabel = false,
  } = props;
  const hasLabel = Boolean(label);
  const containerClassNames = classNames({
    "alt-label": hasLabel && altLabel,
    [className ?? ""]: className,
  });
  const { getDropdownProps, addSelectedItem, removeSelectedItem, selectedItems } =
    useMultipleSelection({
      itemToString: (item) => item.value,
      selectedItems: value ?? [],
      onSelectedItemsChange: ({ selectedItems }) => {
        if (selectedItems) {
          onChange(selectedItems ?? []);
        }
      },
    });
  const {
    isOpen,
    getToggleButtonProps,
    getMenuProps,
    highlightedIndex,
    getItemProps,
    getLabelProps,
    closeMenu,
  } = useSelect({
    items: options,
    circularNavigation: true,
    stateReducer: (_, { changes, type }) => {
      const itemInItems = (selectedItems as MultiSelectOption[]).find(
        ({ value }) => value === changes.selectedItem?.value,
      );

      switch (type) {
        case useSelect.stateChangeTypes.MenuKeyDownEnter:
        case useSelect.stateChangeTypes.MenuKeyDownSpaceButton:
        case useSelect.stateChangeTypes.ItemClick:
          if (itemInItems && changes.selectedItem) {
            removeSelectedItem(changes.selectedItem);
          } else if (changes.selectedItem) {
            addSelectedItem(changes.selectedItem);
          }

          return {
            ...changes,
            isOpen: true, // keep the menu open after selection.
          };
      }
      return changes;
    },
  });

  return (
    <div
      css={(theme): SerializedStyles => multiSelectContainer(theme, { isOpen, block, size })}
      className={containerClassNames}
    >
      {hasLabel && <Label {...getLabelProps({ id })}>{label}</Label>}
      <button type="button" className="select-btn" {...getToggleButtonProps(getDropdownProps())}>
        {(selectedItems.length && (selectedItems[0] as MultiSelectOption).label) || placeholder}
        <CaretDownSVG height="24" />
      </button>
      <ul data-testid="list-container" {...getMenuProps()}>
        <div className="content">
          <Text fontSize="sm"> Select one or more items</Text>
          <button className="close-btn" onClick={(): void => closeMenu()}>
            <DropUpArrowSVG height={20} />
          </button>
        </div>
        {isOpen &&
          options.map((option, index) => {
            const isSelected = selectedItems.find((item) => item.value === option.value);

            return (
              <Option
                key={option.value}
                isHighlighted={highlightedIndex === index}
                isSelected={Boolean(isSelected)}
                {...getItemProps({ item: option, index, "aria-selected": Boolean(isSelected) })}
              >
                {option.label}
              </Option>
            );
          })}
      </ul>
    </div>
  );
};

export default MultiSelect;
