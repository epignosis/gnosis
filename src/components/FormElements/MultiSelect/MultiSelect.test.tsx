import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import MultiSelect from "./MultiSelect";
import { screen, render } from "@test-utils/render";

const OPTIONS = [
  {
    label: faker.company.companyName(),
    value: faker.company.companySuffix(),
  },
  {
    label: faker.commerce.department(),
    value: faker.commerce.productAdjective(),
  },
  {
    label: faker.git.commitMessage(),
    value: faker.git.commitSha(),
  },
  {
    label: faker.hacker.phrase(),
    value: faker.hacker.abbreviation(),
  },
];

describe("<MultiSelect />", () => {
  it("renders correctly", () => {
    const mockOnChange = jest.fn();
    const labelTxt = faker.company.catchPhrase();
    const placeholderTxt = faker.company.catchPhraseDescriptor();
    render(
      <MultiSelect
        label={labelTxt}
        placeholder={placeholderTxt}
        options={OPTIONS}
        onChange={mockOnChange}
      />,
    );

    const label = screen.getByText(labelTxt);
    const input = screen.getByText(placeholderTxt);
    const list = screen.queryByTestId(/list-container/i);

    expect(label).toHaveTextContent(labelTxt);
    expect(input).toBeInTheDocument();
    expect(list).not.toBeVisible();
  });

  it("opens and chooses a value", () => {
    const mockOnChange = jest.fn();
    const labelTxt = faker.company.catchPhrase();
    const placeholderTxt = faker.company.catchPhraseDescriptor();
    render(
      <MultiSelect
        label={labelTxt}
        placeholder={placeholderTxt}
        options={OPTIONS}
        onChange={mockOnChange}
      />,
    );

    const input = screen.getByText(placeholderTxt);
    const list = screen.queryByTestId(/list-container/i);

    userEvent.click(input);

    expect(list).toBeVisible();

    const options = screen.getAllByRole("option");
    const firstOption = options[1];

    expect(options).toHaveLength(OPTIONS.length);
    expect(firstOption).toHaveAttribute("aria-selected", "false");
  });

  it("closes when close button is pressed", () => {
    const mockOnChange = jest.fn();
    const labelTxt = faker.company.catchPhrase();
    const placeholderTxt = faker.company.catchPhraseDescriptor();
    render(
      <MultiSelect
        label={labelTxt}
        placeholder={placeholderTxt}
        options={OPTIONS}
        onChange={mockOnChange}
      />,
    );

    const input = screen.getByText(placeholderTxt);
    const list = screen.queryByTestId(/list-container/i);

    userEvent.click(input);

    expect(list).toBeVisible();

    const closeBtn = screen.getByTestId("close-btn");

    userEvent.click(closeBtn);

    expect(list).not.toBeVisible();
  });

  it("matches snapshot", () => {
    const mockOnChange = jest.fn();
    const { container } = render(
      <MultiSelect
        label="Test Multi-select"
        placeholder="Test placeholder"
        options={OPTIONS}
        onChange={mockOnChange}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with `block = true`", () => {
    const mockOnChange = jest.fn();
    const { container } = render(
      <MultiSelect
        label="Test Multi-select"
        placeholder="Test placeholder"
        options={OPTIONS}
        onChange={mockOnChange}
        block
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with `altLabel = true`", () => {
    const mockOnChange = jest.fn();
    const { container } = render(
      <MultiSelect
        label="Test Multi-select"
        placeholder="Test placeholder"
        options={OPTIONS}
        onChange={mockOnChange}
        altLabel
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
