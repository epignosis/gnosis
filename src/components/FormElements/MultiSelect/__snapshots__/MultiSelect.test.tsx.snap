// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelect /> matches snapshot 1`] = `
<div>
  <div
    class=" css-kh89cj-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-multiSelectContainer-multiSelectContainer"
  >
    <label
      class="css-xs6o3g-label-label"
      for="downshift-6-toggle-button"
    >
      Test Multi-select
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-6-label downshift-6-toggle-button"
      class="select-btn"
      id="downshift-6-toggle-button"
      type="button"
    >
      Test placeholder
      <svg-file-stub
        height="24"
      />
    </button>
    <ul
      aria-labelledby="downshift-6-label"
      data-testid="list-container"
      id="downshift-6-menu"
      role="listbox"
      tabindex="-1"
    >
      <div
        class="content"
      >
        <span
          class="css-26eo9i-textContainer-textContainer"
          data-testid="text-component"
        >
           Select one or more items
        </span>
        <button
          class="close-btn"
          data-testid="close-btn"
        >
          <svg-file-stub
            height="20"
          />
        </button>
      </div>
    </ul>
  </div>
</div>
`;

exports[`<MultiSelect /> matches snapshot with \`altLabel = true\` 1`] = `
<div>
  <div
    class="alt-label css-kh89cj-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-multiSelectContainer-multiSelectContainer"
  >
    <label
      class="css-xs6o3g-label-label"
      for="downshift-8-toggle-button"
    >
      Test Multi-select
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-8-label downshift-8-toggle-button"
      class="select-btn"
      id="downshift-8-toggle-button"
      type="button"
    >
      Test placeholder
      <svg-file-stub
        height="24"
      />
    </button>
    <ul
      aria-labelledby="downshift-8-label"
      data-testid="list-container"
      id="downshift-8-menu"
      role="listbox"
      tabindex="-1"
    >
      <div
        class="content"
      >
        <span
          class="css-26eo9i-textContainer-textContainer"
          data-testid="text-component"
        >
           Select one or more items
        </span>
        <button
          class="close-btn"
          data-testid="close-btn"
        >
          <svg-file-stub
            height="20"
          />
        </button>
      </div>
    </ul>
  </div>
</div>
`;

exports[`<MultiSelect /> matches snapshot with \`block = true\` 1`] = `
<div>
  <div
    class=" css-g1aj0v-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-multiSelectContainer-multiSelectContainer"
  >
    <label
      class="css-xs6o3g-label-label"
      for="downshift-7-toggle-button"
    >
      Test Multi-select
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-7-label downshift-7-toggle-button"
      class="select-btn"
      id="downshift-7-toggle-button"
      type="button"
    >
      Test placeholder
      <svg-file-stub
        height="24"
      />
    </button>
    <ul
      aria-labelledby="downshift-7-label"
      data-testid="list-container"
      id="downshift-7-menu"
      role="listbox"
      tabindex="-1"
    >
      <div
        class="content"
      >
        <span
          class="css-26eo9i-textContainer-textContainer"
          data-testid="text-component"
        >
           Select one or more items
        </span>
        <button
          class="close-btn"
          data-testid="close-btn"
        >
          <svg-file-stub
            height="20"
          />
        </button>
      </div>
    </ul>
  </div>
</div>
`;
